include/SFML/Graphics/VertexBuffer.hpp

```diff
@@ -0,0 +1,389 @@
+////////////////////////////////////////////////////////////
+//
+// SFML - Simple and Fast Multimedia Library
+// Copyright (C) 2007-2017 Laurent Gomila (laurent@sfml-dev.org)
+//
+// This software is provided 'as-is', without any express or implied warranty.
+// In no event will the authors be held liable for any damages arising from the use of this software.
+//
+// Permission is granted to anyone to use this software for any purpose,
+// including commercial applications, and to alter it and redistribute it freely,
+// subject to the following restrictions:
+//
+// 1. The origin of this software must not be misrepresented;
+//    you must not claim that you wrote the original software.
+//    If you use this software in a product, an acknowledgment
+//    in the product documentation would be appreciated but is not required.
+//
+// 2. Altered source versions must be plainly marked as such,
+//    and must not be misrepresented as being the original software.
+//
+// 3. This notice may not be removed or altered from any source distribution.
+//
+////////////////////////////////////////////////////////////
+
+#ifndef SFML_VERTEXBUFFER_HPP
+#define SFML_VERTEXBUFFER_HPP
+
+////////////////////////////////////////////////////////////
+// Headers
+////////////////////////////////////////////////////////////
+#include <SFML/Graphics/Export.hpp>
+#include <SFML/Graphics/PrimitiveType.hpp>
+#include <SFML/Graphics/Drawable.hpp>
+#include <SFML/Window/GlResource.hpp>
+
+
+namespace sf
+{
+class RenderTarget;
+class Vertex;
+
+////////////////////////////////////////////////////////////
+/// \brief Vertex buffer storage for one or more 2D primitives
+///
+////////////////////////////////////////////////////////////
+class SFML_GRAPHICS_API VertexBuffer : public Drawable, private GlResource
+{
+public:
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Usage specifiers
+    ///
+    ////////////////////////////////////////////////////////////
+    enum Usage
+    {
+        Stream,  ///< Constantly changing data
+        Dynamic, ///< Occasionally changing data
```

Comment:
I really wonder what difference this flag (`Dynamic`) makes at driver level. I can understand that choosing between constantly changing or static data can make a difference, but "occasionally changing"? How do you optimize that?

Well, my point is, are there use cases where using `Dynamic` gives a significant performance boost compared to the 2 other options, in the context of SFML? Or is it just for the sake of completeness, given that this class is a VBO wrapper that can be used with OpenGL code?

Reply:
Considering D3D also has several usage flags (see [here](https://msdn.microsoft.com/en-us/library/windows/desktop/ff476259(v=vs.85).aspx)), I take it it actually makes a difference somehow. At the end of the day, this is a usage _hint_ anyway. Depending on the driver implementation it can choose to trust us or doubt us. We aren't going to be doing it a favour by making it choose a wrong starting point. Also see chapter 28.2.2 of [this](https://www.seas.upenn.edu/~pcozzi/OpenGLInsights/OpenGLInsights-AsynchronousBufferTransfers.pdf).